### ISSUE TITLE
Using SpanBERT to predict spans

### ISSUE THREAD

**Issue by sam-qordoba** [facebookresearch/SpanBERT#29]
Hi, I maintain [FitBERT](https://github.com/Qordobacode/fitbert), a library that uses BERT to fill in the blanks (do masked language modeling). An example:

```python
from fitbert import FitBert


# in theory you can pass a model_name and tokenizer
# currently supported models: bert-large-uncased and distilbert-base-uncased
# this takes a while and loads a whole big BERT into memory
fb = FitBert()

masked_string = "Why Bert, you're looking ***mask*** today!"
options = ['buff', 'handsome', 'strong']

ranked_options = fb.rank(masked_string, options=options)
# >>> ['handsome', 'strong', 'buff']
```

I think that SpanBERT should be an improvement to using BERT when the mask covers more than 1 token. However, I am not seeing SBO anywhere in this code. Is it not released? Should I just mask sequential tokens and hope that the model is better at predicting them?

Also, since you don't do NSP, does the `[SEP]` tag have any meaning? If not, is there no special EOS token?

Thanks so much!

**Comment by mandarjoshi90** [comment]
Hi Sam,
1. The pre-training code hasn't been released yet. Masking contiguous chunks of tokens should still work well for your case.
2. CLS and SEP are still being used in the input during pretraining, but as such the model could very well have been trained without them. It's convenient to include them since quite a few sentence classification tasks use the CLS token.

**Comment by sam-qordoba** [comment]
Thanks @mandarjoshi90 !

**Comment by chrisjbryant** [comment]
Hey Sam,

I'm wondering if you managed to get span prediction to work. I tried loading the pre-trained models as a masked LM in huggingface, but the predictions are random and don't make any sense.

I also get the following warning when I load the model:
`INFO:transformers.modeling_utils:Weights of BertForMaskedLM not initialized from pretrained model: ['bert.pooler.dense.weight', 'bert.pooler.dense.bias', 'cls.predictions.bias', 'cls.predictions.transform.dense.weight', 'cls.predictions.transform.dense.bias', 'cls.predictions.transform.LayerNorm.weight', 'cls.predictions.transform.LayerNorm.bias', 'cls.predictions.decoder.weight', 'cls.predictions.decoder.bias']`

I'm now confused whether the released models are pre-trained for a different task or just pre-configured and require fine-tuning.

Any help appreciated!

**Comment by sam-qordoba** [comment]
Hey Chris!

Fancy running into you here...

I did get span prediction to work, but I haven't yet evaluated if it is indeed working better than a regular BERT model.

Here is what I did (in a notebook, so `!` indicates a shell command):

```jupyter
from transformers import BertForMaskedLM, BertTokenizer

!wget https://dl.fbaipublicfiles.com/fairseq/models/spanbert_hf_base.tar.gz

!tar -xzvf spanbert_hf_base.tar.gz

bert = BertForMaskedLM.from_pretrained('/content/')  # have to pass the directory!!! ARG

tok_path = 'bert-base-cased' # uses same tokenizer in bert/spanbert

tokenizer = BertTokenizer.from_pretrained(tok_path)
```

I did not see the warning you get, but this was over a month ago, so who knows if it still works... :grimacing:

**Comment by chrisjbryant** [comment]
Haha, yeah, I think we're definitely following similar lines at the moment. ;)

I'm doing the same as above (thanks to your other reply on how to get it to work in huggingface!), but get useless predictions. 

Example text: "I like to [MASK] books."  
`bert-base-cased` output:
```
read tensor(0.7716, device='cuda:1')
do tensor(0.0226, device='cuda:1')
write tensor(0.0202, device='cuda:1')
```
Spanbert output:
```
##uad tensor(0.0191, device='cuda:1')
##dial tensor(0.0114, device='cuda:1')
##iq tensor(0.0048, device='cuda:1')
```

I get a different warning when I load standard BERT too:
```
INFO:transformers.modeling_utils:Weights of BertForMaskedLM not initialized from pretrained model: ['cls.predictions.decoder.bias']
INFO:transformers.modeling_utils:Weights from pretrained model not used in BertForMaskedLM: ['cls.seq_relationship.weight', 'cls.seq_relationship.bias']
```

I can only conclude it's something to do with the weights not loading that's the problem. The rest of my code is exactly the same otherwise. 

I did also test this with an older version of huggingface, just in case that was the issue, but no joy...

**Comment by sam-qordoba** [comment]
@chrisjbryant Ah, so the SpanBERT pretrained model is specifically `BertForMultipleChoice`?

**Comment by chrisjbryant** [comment]
Or `BertForSomethingElse`. 

Based on the last post in that thread, it looks like they pretrained a model with a masked LM head, but then replaced it with a task-specific head. So unless you have the original pretrained model, you can't use the masked LM. Hopefully this is something SpanBERT authors can provide! &ast;poke&ast;

**Comment by mandarjoshi90** [comment]
Hi folks,
I'll try provide the masked LM head soon. I'm no longer at FB which means releasing non-standard stuff takes time. I hope to get back sometime next week. Thanks for your patience.

**Comment by vnagapudi** [comment]
Hi Mandar,
Did you get to the masked LM by any chance? 
Thanks
Venky

**Comment by mandarjoshi90** [comment]
Unfortunately, not. Still waiting. Sorry.

**Comment by SkanderHellal** [comment]
Hello everyone, I am asking if fitbert can solve the problem of filling the blanks with multiple words.For example, if I a masking the work pizza in the sentence I am eating pizza with friends, is it possible to fill the blank with delicious food or any other contiguous chunk of words, the same if I am masking a contiguous chunk of words is it possible to replace it with one word (The same example but replacing delicious food with pizza)

**Comment by sam-qordoba** [comment]
> Hello everyone, I am asking if fitbert can solve the problem of filling the blanks with multiple words.For example, if I a masking the work pizza in the sentence I am eating pizza with friends, is it possible to fill the blank with delicious food or any other contiguous chunk of words, the same if I am masking a contiguous chunk of words is it possible to replace it with one word (The same example but replacing delicious food with pizza)

This is probably not the place to discuss FitBERT... here is the issue for doing that: https://github.com/Qordobacode/fitbert/issues/19

There's a link with a notebook showing how to hack it

**Comment by gchronis** [comment]
Hello! I am also very much interested in using SpanBERT for MLM. Is this an update that is still in the works? I highly doubt my own capacity to get this working, but would the maintainers of the codebase accept an external PR which adds this functionality?

**Comment by chrisjbryant** [comment]
Hi @gchronis . You might like to take a look at #38 as I managed to use the download link there to get SpanBERT working as a MLM. Hope that helps!
